modules/module-mysql/src/replication/zongji/BinLogListener.ts

```diff
@@ -0,0 +1,206 @@
+  onDelete: (rows: Row[], tableMap: TableMapEntry) => Promise<void>;
+  onCommit: (lsn: string) => Promise<void>;
+}
+
+export interface BinLogListenerOptions {
+  connectionManager: MySQLConnectionManager;
+  eventHandler: BinLogEventHandler;
+  includedTables: string[];
+  serverId: number;
+  startPosition: common.BinLogPosition;
+}
+
+/**
+ *  Wrapper class for the Zongji BinLog listener. Internally handles the creation and management of the listener and posts
+ *  events on the provided BinLogEventHandler.
+ */
+export class BinLogListener {
+  private connectionManager: MySQLConnectionManager;
+  private eventHandler: BinLogEventHandler;
+  private binLogPosition: common.BinLogPosition;
+  private currentGTID: common.ReplicatedGTID | null;
+
+  zongji: ZongJi;
+  processingQueue: async.QueueObject<BinLogEvent>;
+
+  constructor(public options: BinLogListenerOptions) {
+    this.connectionManager = options.connectionManager;
+    this.eventHandler = options.eventHandler;
+    this.binLogPosition = options.startPosition;
+    this.currentGTID = null;
+
+    this.processingQueue = async.queue(this.createQueueWorker(), 1);
+    this.zongji = this.createZongjiListener();
+  }
+
+  public async start(): Promise<void> {
+    if (this.isStopped) {
+      return;
+    }
+    logger.info(`Starting replication. Created replica client with serverId:${this.options.serverId}`);
+
+    this.zongji.start({
+      // We ignore the unknown/heartbeat event since it currently serves no purpose other than to keep the connection alive
+      // tablemap events always need to be included for the other row events to work
+      includeEvents: ['tablemap', 'writerows', 'updaterows', 'deleterows', 'xid', 'rotate', 'gtidlog'],
+      includeSchema: { [this.connectionManager.databaseName]: this.options.includedTables },
+      filename: this.binLogPosition.filename,
+      position: this.binLogPosition.offset,
+      serverId: this.options.serverId
+    } satisfies StartOptions);
+
+    return new Promise<void>((resolve, reject) => {
+      this.zongji.on('error', (error) => {
+        if (!this.isStopped) {
+          logger.error('Binlog listener error:', error);
+          this.stop();
+          reject(error);
+        } else {
+          logger.warn('Binlog listener error during shutdown:', error);
+        }
+      });
+
+      this.processingQueue.error((error) => {
+        if (!this.isStopped) {
+          logger.error('BinlogEvent processing error:', error);
+          this.stop();
+          reject(error);
+        } else {
+          logger.warn('BinlogEvent processing error during shutdown:', error);
+        }
+      });
+
+      this.zongji.on('stopped', () => {
+        resolve();
+        logger.info('BinLog listener stopped. Replication ended.');
+      });
+
+      // Handle the edge case where the listener has already been stopped before completing startup
+      if (this.isStopped) {
```

Comment:
Just some thoughts/questions from my side on the logic here.

Should we remove the listeners which have been added here with `this.zongji.on` when we complete this promise? Or are they configured to only run once?

I also assume the logic in this promise creator callback is synchronous, which makes me wonder if we should be checking `this.isStopped` before registering the event handlers? Is it possible that we could trigger a stop event while creating this promise?